### ISSUE TITLE
Convert htmx to TypeScript 😍

### ISSUE THREAD

**Issue by bencroker** [bigskysoftware/htmx#606]
TypeScript would bring lots of DX benefits to the htmx source code. Besides the obvious strict type checking, we would be able to write code using the latest and greatest ES2020, but continue compiling to ES5 (to support IE11), or any version of ECMAScript really, with a single config setting.

```json
{
  "compilerOptions": {
    "target": "ES5"
  }
}
```

I've started a WIP branch in which I've converted `htmx.js` to `htmx.ts`, changing the minimum amount of code possible, and it is compiling with all tests passing. 

@1cg in theory, we could merge this branch into the `dev` branch before the next minor release, and start converting the source code incrementally.

**Comment by gnat** [comment]
Of course up to @1cg to decide if the benefits are actually worthwhile but personally would vote against moving over to TS:

* Increased barrier for contributions.
* More complex, slower build process.
* The true "futureproof" choice is to continue evolving with the latest ECMAScript standard. (I'd rather wait to have official types support in ECMAScript.)

**Comment by gnat** [comment]
Of course up to @1cg to decide if this switch is actually worthwhile, but will have to say no thanks because:

* Increased barrier for contributions.
* More complex, slower build process.
* Type checking may not even be of tangible benefit for htmx (htmx has a nice test suite).

The ECMAScript spec will probably supersede TypeScript in the near future anyway. This has happened historically with other dead Javascript dialects (example: when classes were included in ES6).

**Comment by bencroker** [comment]
While you may have some good points, I still think that the benefits of moving to a modern codebase (rather than one that resembles 2014, no offence @1cg!) far outweigh the arguments given, especially for the core contributors of htmx.

> Increased barrier for contributions.

That's debatable, as TypeScript is a superset of JavaScript and helps to catch errors early.

> More complex, slower build process.

The build process remains the same: `npm run dist`

> Type checking may not even be of tangible benefit for htmx (htmx has a nice test suite).

Type checking helps to write more robust code and leads to tighter integration with code editors that can then catch errors much more reliably. This doesn't question the value of the test suite at all!

**Comment by khalwat** [comment]
In general, I think any library should be written in TypeScript... and probably everything else, too, for the reasons cited above.

> The ECMAScript spec will probably supersede TypeScript in the near future anyway. This has happened historically with other dead Javascript dialects (example: when classes were included in ES6).

There's just so much momentum behind TypeScript (and Microsoft as well) that I don't see TypeScript going anywhere any time soon. So I don't think we're looking at CoffeeScript all over again.

Remember, TypeScript just _adds_ types to JavaScript. Strip the types, and you have fully up to spec JavaScript code.

If anything, I could see TypeScript being subsumed into JavaScript at some point... which means your code would likely still work

But let's take the worst-case, which is that in 4-5 years (given how slowly these standards get implemented, I think this is likely reasonable), ECMAscript has some kind of incompatible type system built into it. You'll still have benefitted for 4-5 years from all the things TypeScript gives you, and your code will still compile down to JavaScript regardless of any standards change.

And it's very likely even if/when types do come to the ECMAscript standard that you'll be transpiling for years after that anyway, for backward compatibility reasons.

**Comment by khalwat** [comment]
> More complex, slower build process

`tsup` is what I'd recommend for builds -- it's actually quite fast and simple:

https://github.com/egoist/tsup

I've used it before in some plugins I've written, and the implementation is easy:

https://github.com/nystudio107/rollup-plugin-critical/blob/master/package.json#L30

**Comment by bencroker** [comment]
Here's my WIP fork: https://github.com/bencroker/htmx/tree/typescript

**Comment by leevigraham** [comment]
Here's my type definitions: https://github.com/leevigraham/typescript-htmx-types

**Comment by khalwat** [comment]
@leevigraham should submit them to Definitely Typed -> https://github.com/DefinitelyTyped/DefinitelyTyped

Nice work! Even as a WIP it's useful!

**Comment by leevigraham** [comment]
@khalwat That was the plan I think but I was just learning typescript so I wasn't sure it was up to snuff. Then I got the very strong impression that there was no future for TS in HTMX so I kinda lost interest.

**Comment by gnat** [comment]
> The ECMAScript spec will probably supersede TypeScript in the near future anyway. This has happened historically with other dead Javascript dialects (example: when classes were included in ES6).

https://www.youtube.com/watch?v=O9F4K804XC8

History repeats itself.

RIP Typescript soon.

Also re-removing the build step = Finally :+1:  @1cg

**Comment by khalwat** [comment]
Firstly, I'm thrilled if typing becomes part of the actual ECMAScript spec. That's an ideal world scenario for me.

Arguably, if things like TypeScript didn't exist, the official spec would not be pushed forward. I think it's great that various JS technologies are rolled out, and the ones that get proven "in the field" are rolled into the spec.

This cycle will continue into the future; there will likely always be a build step to allow for the use of modern versions of the language, whilst still offering backward compatibility.

It takes a long time for the ecosystem to catch up. The spec has to be drawn up, approved, and then browser vendors slowly roll it out with varying degrees of support.

So the idea of "no build step" probably will never happen, if you want to be using technologies as they evolve.

**Comment by adrienne** [comment]
I would like to add a +1 for converting to Typescript, or at least adding a definitions file.

**Comment by bencroker** [comment]
@adrienne there's one at https://github.com/bigskysoftware/htmx/blob/master/src/htmx.d.ts

**Comment by adrienne** [comment]
@bencroker is there a quick demo anywhere of importing/making HTMX work nicely? (I'm still learning TS.)

**Comment by bencroker** [comment]
I'm not sure, the [Discord server](https://htmx.org/discord) would be the place to ask.

**Comment by leevigraham** [comment]
@bencroker Guess I can archive my repo then.

**Comment by bencroker** [comment]
@leevigraham I haven't compared them, feel free to PR any missing definitions to this repo.

**Comment by thomaseizinger** [comment]
> Type checking may not even be of tangible benefit for htmx (htmx has a nice test suite).

Here is a recent example of a bug that would have been caught be a type checker and wasn't caught by the existing test suite: https://github.com/bigskysoftware/htmx/pull/1156

A nice test suite doesn't mean you wouldn't benefit from types. On the contrary, types are a form of tests and running a type checker is like running a test suite. When applied well, types can supersede a whole class of tests like "Am I forwarding the arguments in the right order".

**Comment by benpate** [comment]
Yeah, I'm a big fan of Typescript, too.  But it's unlikely that htmx will get translated fully into TS.  

Fortunately, we have made the compromise of using Typescript/JSDoc comments, which are a little more bulky but still 100% Javascript compatible ([see my note above](https://github.com/bigskysoftware/htmx/issues/606#issuecomment-942878363))

Since I can still use the TS compiler to statically analyze the htmx source code, I've found that using JSDoc comments  gets me about 95% of the benefits of using Typescript without having a separate compile step.

Does this "comments only" approach work for you, @thomaseizinger?  We've made some progress in the past towards adding accurate JSDoc annotations and a Typescript `.d.ts` header file.  Do you have any interest in making improvements there?

**Comment by khalwat** [comment]
JSDoc is documentation oriented, while Typescript adds static typing to Javascript. These tools do have some parts in common, they allow for type checks, for example, but their purpose is different.

Also, remember that if you strip the types from Typescript, you end up with valid Javascript, that's baked into the spec of the language. So I'd call it a build step, not a compile step.

While adding documentation is always good, you can have both: Typescript and also TSDoc, they aren't mutually exclusive. For any library, I think Typescript is a really good choice.

You'll be surprised at how many bugs you find and fix in the process of refactoring it to Typescript, and you can do it incrementally, slowly making it more and more strict as you refactor.

**Comment by benpate** [comment]
@khalwat - Yes.  I'm very familiar with Typescript, and I'd love to use Typescript in htmx, but that's not an option.

However, [the Typescript compiler CAN parse JS files directly](https://www.typescriptlang.org/docs/handbook/intro-to-js-ts.html), and [Typescript uses JSDoc comments directly](https://www.typescriptlang.org/docs/handbook/jsdoc-supported-types.html).  It's not a separate toolchain; it's another way into the Typescript ecosystem.

As I said above, JSDoc comments are not perfect -- they takes up more space and can be a little clumsy in places -- but using this style allows us to have 95% of the benefits of Typescript WITHOUT introducing a separate compile dependency, which is the main requirement that prevents us from using Typescript outright.

It's great to have more fellow Typescript zealots in the htmx community.  If you're interested, check out what we've got so far and in the way of `*.d.ts` headers and JSDoc comments.  Pull requests that shore up internal consistency and reliability will be a no-brainer to merge :)

**Comment by khalwat** [comment]
@benpate Yeah those comments were also intended for anyone else reading, just to clarify things.

Maybe I'm missing why htmx can't be refactored to Typescript? The main issue seems to be the build step? Can you help me understand why this is a blocker?

 If htmx is published as an npm package, you can add a `prepublishOnly` script that does the build transparently, and `files` to designate only the `dist/` (or whatever) files should be published via npm, which should make everything pretty seamless.

**Comment by thomaseizinger** [comment]
If that is the preferred workflow, I think it would make sense to run that as part of CI to make sure it doesn't get outdated!

What is the reasoning that a separate compile dependency is a big problem? There is already minification and testing as part of the build/publish workflow so it doesn't sound like a heavy addition?

I very much enjoy that as a user of HTMX, I don't need a build pipeline at all but if one exists here already, extending doesn't seem that bad? Perhaps I am missing something!

**Comment by khalwat** [comment]
> What is the reasoning that a separate compile dependency is a big problem? There is already minification and testing as part of the build/publish workflow so it doesn't sound like a heavy addition?
> 
> I very much enjoy that as a user of HTMX, I don't need a build pipeline at all but if one exists here already, extending doesn't seem that bad? Perhaps I am missing something!

Currently there's a `dist` script that looks like this:

```js
    "dist": "cp -r src/* dist/ && npm run-script uglify && gzip -9 -k -f dist/htmx.min.js > dist/htmx.min.js.gz && exit",
```

Ironically, it actually could be simplified when converted to Typescript using [`tsup`](https://github.com/egoist/tsup) to something like [this](https://github.com/nystudio107/rollup-plugin-critical/blob/master/package.json#L30):

```
    "build": "tsup src/index.ts --sourcemap --dts --format cjs,esm",
```

...and then to make things easier, just add a `prepublishOnly` script that automatically runs `build` whenever a new version is published.

As you said, there's already a build step, I don't see that as being an issue.

**Comment by 1cg** [comment]
i like being able to debug the unmodified version of javascript locally when I develop, rather than having to have a build step or source maps or whatever.

JSDoc has been a half-way house giving us some type checking (I'm an old java developer, I am not opposed to types) which I like.

Do browsers support typescript directly yet?

**Comment by khalwat** [comment]
No browsers I'm aware of support Typescript directly. I think you'd be surprised by the DX benefits you get from something like Typescript in combination with Vite for HMR when developing.

You get the static type checking benefits from Typescript, with hot module replacement of the code while developing, so never having to hit the refresh button in your browser while developing.

Incidentally, you can use Deno if you want a Node-like environment that natively supports Typescript.

**Comment by thomaseizinger** [comment]
Another benefit would be that you can make use of the module system and split up the current monolith into independent modules.

**Comment by khalwat** [comment]
yep... and tree shaking will remove unused code... just lots of benefits to using a modern dev system imo

**Comment by gnat** [comment]
htmx has excellent long term staying power because of it's simplicity, zero dependencies, ultra-low friction to contribute or debug.

This has always been a killer feature of htmx

It's why we're here and not just using React / Svelte / etc. htmx is a uniquely magestic monolith in this space with a lot of passionate contributors.

On the other hand, I realise it's tempting and easy to add extra dependencies and build steps using an "in vogue" build system of the month, but this stuff is difficult or impossible to remove in the future, and will put htmx on a dependency death clock at worst- Like most other libraries in this space.

Not to mention the "developer judgement" (lol) associated with "having X build system" when it goes out of fashion. :roll_eyes: It'll happen.

...

> i like being able to debug the unmodified version of javascript locally when I develop, rather than having to have a build step or source maps or whatever.

It's a highly refreshing experience.

**Comment by khalwat** [comment]
> htmx has excellent long term staying power because of it's simplicity, zero dependencies, ultra-low friction to contribute or debug.

All of which would be greatly enhanced by a modern build chain.

Anyway, I give up -- carry on.

**Comment by thomaseizinger** [comment]
> htmx has excellent long term staying power because of it's simplicity, zero dependencies, ultra-low friction to contribute or debug.

Allow me to weigh in here as a first time user and first time contributor.

**User perspective:** I love being able to use it by just referencing the CDN-version in my HTML. It is a massive plus for me.

**Contributor perspective** (https://github.com/bigskysoftware/htmx/pull/1156):

- Debugging proved difficult until I discovered that there is also a non-minified version on the CDN. It is hard to set a break point on a single line of code.
- `npm install` initially failed because `node-sass` still depends on python2 which I banned from my computer. After quickly checking what is used for, I decided to comment out the dependency and move on because I wouldn't need it for debugging my issue.
- Navigating the source code proved difficult because the file is very large and no types means you don't know what is being passed around. Fortunately, the Jetbrains IDEs have reasonable good JS integration, so navigating between functions was okay.
- After I found the bug, my first reaction was: "This would have been caught be a type system.", which lead me to this issue.
- Writing a test was easy to start with. I enjoyed the existence of certain utilities.
- Running the test was a bit of an adventure. I had made a mistake in where I placed the `hx-boost` attribute (turns out it only works on child elements, not on the element it is being applied to) which caused the test page to refresh constantly (the browser actually submitted the POST request). It took me a while to figure out why the test ran in an infinite loop. I followed the README which said to run the tests in the browser. After checking now, I see that there is also `npm test` which may have made this easier.

There are always learnings when you contribute to a new project because everyone does things a little differently and that is okay. My next contribution to HTMX is likely going to be easier, now that I know some specifics of the project.

Overall, I'd say it wasn't particularly difficult to contribute. I got there in the end and managed to fix the bug but I am struggling to agree to it being "ultra-low friction".

**Comment by bencroker** [comment]
With development of htmx v2 planned to start in the coming weeks/months, I thought I'd post a few updates and related thoughts. 

1. According to the State of JS 2022 survey results, significantly [more people are writing more TypeScript than vanilla JavaScript](https://2022.stateofjs.com/en-US/usage/#js_ts_balance).

<img width="1188" alt="Screenshot 2023-01-18 at 10 20 20" src="https://user-images.githubusercontent.com/57572400/213233540-1f1ce671-839a-4d59-ac52-3762caa64325.png">

htmx also [made the list this year](https://2022.stateofjs.com/en-US/libraries/front-end-frameworks/#front_end_frameworks_others)!!

<img width="1193" alt="Screenshot 2023-01-18 at 10 36 54" src="https://user-images.githubusercontent.com/57572400/213240215-0374f9cd-cec1-421c-a5e9-1616da7dc0c4.png">

2. Browser support for htmx v2 is yet to be decided. Supporting IE11 and being limited to writing ES5 only has always felt like a thorn in my side when contributing to htmx until now. With TypeScript, we can write code using the latest and greatest ECMAScript and compile to ES6 or any version of ECMAScript (even multiple distribution versions, if desirable).

3. The htmx source code has naturally grown over time. Having a codebase consisting of a [single file with 3,399 lines of code](https://github.com/bigskysoftware/htmx/blob/v1.8.5/src/htmx.js) feels like an antiquated approach. Using TypeScript, or any more modern approach, would allow the code to be componentised, better documented, and therefore more approachable to contributors.

**Comment by gnat** [comment]
Github counts typescript comments, adding indentation, etc. all equally to core code contributions :roll_eyes: I don't think it's entirely fair to re-architect htmx in a direction that will add substantially more files / code / build complexity, because of a few contributions.

The current idea of including separate typescript definitions is a great direction because it satisfies specific developers without blowing up the files / code / build complexity.. etc. for every other developer involved.

The strictness of CORS (you'll require a web server to run your javascript) is what prevents us from moving to full-on javascript modules at this time- it's not actually IE11.

**Comment by khalwat** [comment]
@gnat a local dev setup like Vite, in addition to making the development DX significantly better with hot module replacement, will also provide a web server to run your Javascript.

Oh, and it does TypeScript out of the box. Win-win.

**Comment by leevigraham** [comment]
@bencroker Maybe a HTMXTS fork?

**Comment by leevigraham** [comment]
@bencroker Hopefully there will be some solid JSDoc comments which will fill the gap.

**Comment by khalwat** [comment]
> @bencroker Hopefully there will be some solid JSDoc comments which will fill the gap.

Won't quite fill the gap _completely_, but I agree it's better than no JSDoc comments. Good article:

https://blog.logrocket.com/typescript-vs-jsdoc-javascript/

Ironically, it'll likely be TypeScript that does the JSDoc type checking regardless.

**Comment by gnat** [comment]
It's worth noting Svelte just abandoned Typescript in favour of JSDoc: https://www.youtube.com/watch?v=zPOHY-cZ1wE

**Comment by khalwat** [comment]
TL;DR Rich wants to be able to monkey patch his libraries.

All I can say is that as someone who went through the ongoing transition of PHP from dynamically typed to statically typed, and have been using Php DocBlocks for typing alongside actual types in PHP as they've been added over the years... the experience of a comment-based type system is not the same.

In any event, hopefully JavaScript will adopt typing in the language itself, and then several years after that, we can all start using it.

**Comment by andryyy** [comment]
Every project reaches the point where things are getting overcomplicated by adopting to X, Y, and Z.

More often than not you gain nothing but more code and workarounds to not lose compatibility for a small or no benefit at all.

New components and workarounds need to be maintained and eventually break at some point leading to frustration and lost time for productive coding on your actual application.

The worst part is adding dependencies on components you don’t control. It kills projects.  

This does not apply to all code obviously. But this right here is a great example for it. 

Another classic is overly complicated CI and automation, that you spend more time maintaining, controlling, and fixing with than anything else.

Sometimes you should stick to the essentials. And never base your decisions on what other people are using or praising.

**Comment by khalwat** [comment]
I think you're vastly overstating the complexity involved here, but I agree that you should never base your decisions on what other people are praising.

You should base your decisions on the value that a particular layer of toolings brings vs. the complexity it adds.

Sounds like the owner of this project has reached that decision, so away we go.

**Comment by unoexperto** [comment]
I'm a backed engineer and lack of TS (and other node shit) is precisely the reason I love HTMX.

**Comment by khalwat** [comment]
The reason to use TypeScript is not "because other people are using it", it's because it can help make your application better with strict & static typing, which will catch errors in your editor rather than at runtime.

It's the same reason why PHP is moving towards stricter typing.

It's also important to keep in mind that TypeScript is literally just JavaScript with types. If you strip the types out of TypeScript code, it is executable JavaScript code.

Anyway, hopefully it (or something like it) becomes part of the ES spec, and we can see the benefits without the consternation!

**Comment by 1cg** [comment]
heya folks, i don't have much to add here beyond saying that I appreciate both perspectives on this and if typescript was native to the browser I would much more strongly consider it

**Comment by bencroker** [comment]
Just dropping a link to this very well written [essay](https://htmx.org/essays/#complexity-very-very-bad) by @alexpetros that was just published:
https://htmx.org/essays/no-build-step/

**Comment by courtesysoft** [comment]
Absolutely do not want typescript in HTMX and am in agreement with gnat.

**Comment by courtesysoft** [comment]
> It's the same reason why PHP is moving towards stricter typing.

PHP is not making typing mandatory though. It is adding typing features, and has had optional typing since 7.2ish ( 2017 ).  Despite this, uptake is still small.

Even better, they're not introducing a compile step like TS does.

The problem with adding strict typing to a dynamic language is that everything that talks to the code is now subject to new rules. This can break dynamically typed code and require a complete refactor. In essence, it has an externality of forcing how other programmers program.

It also adds a dependency to a product maintained by Microsoft, which history has shown is often enough a bad move.

No typescript please. It's not worth the pitfalls.

**Comment by joshmanders** [comment]
> I'm a backed engineer and lack of TS (and other node shit) is precisely the reason I love HTMX.

I'd just like to clarify for some people coming in reading these comments, the usage of TypeScript will only be for the maintainers of the library, not end users. You'd still use htmx the way it is currently being used.

**Comment by alexpetros** [comment]
Much internet ink has been spilled on this settled issue 🔒
